### DISCUSSION TITLE
MooseFS 4.56.5

### DISCUSSION THREAD

**Discussion by pkonopelko** [moosefs/moosefs#584] [Releases]
Dear MooseFS Users,

We are thrilled to announce the release of the long-awaited MooseFS 4.x Community Edition! 🎉

### New Format of Chunk Storage: Erasure Codes (EC)
We have introduced a new format of chunk storage with Erasure Codes, which can store data in ARCHIVE and TRASH states as 4 or 8 parts with an additional parity part, allowing data restoration in case a part is lost. This method provides data restoration in case of missing parts while optimizing storage space:

 - `EC8+1`: Uses 112.5% of space,
 - `EC4+1`: Uses 125% of space,

while maintaining the same level of hardware failure protection as standard replication (200%).

### New Options in Storage Class Definitions

- **`TRASH` State**: files in the MooseFS TRASH, so the deleted files that are retained for a configurable period of time, can now have custom storage definitions within a storage class.
- **`ARCHIVE` State Configuration**: `ARCHIVE` state is now fully configurable:
 - Set file archive times based on last `atime`, `mtime`, or `ctime`.
 - Choose between `FAST` archival (ASAP) or `PER CHUNK`  based on `mtime`.
 - Exclude files below a certain length from being archived.
 - **Logical Expressions in Label Definitions**: Support for comprehensive logical expressions in label definitions is now available, allowing users to create more flexible and precise rules, providing greater control over how storage classes and states are applied to files. This includes the ability to use negations.
 - State-Specific Label Overrides: Override label modes (`STD`, `LOOSE`, `STRICT`) for specific states: `CREATE`, `KEEP`, `ARCHIVE`, or `TRASH`.

### Storage Class Based on File Patterns
Users can now assign file storage classes dynamically based on custom file name patterns, providing greater flexibility in managing data. For example, files with a `.backup` extension could be assigned to a specific storage class with lower redundancy, while files starting with `important_` might be placed in a high-redundancy class.

### Broadcast Mechanism for Server Restarts
Server restarts (both chunk server and master) no longer block I/O until the end of chunk registration process.

### Improved Chunk Format
Enhanced handling of control sums for more secure chunk storage.

### New Web-based Graphical User Interface with Dark Mode
Additional monitoring charts and system insights, offering improved chart systems and more detailed information. Moreover, dark mode is now available.

### Multi-LAN Configuration
Support for clients across different LANs to remain within their LAN IP class while communicating with the master server. Can be used alongside the `remap` option known from MooseFS 3.x.

### Enhanced IP Remap
Customizable IP-to-IP remapping instead of just fixed remapping of leading bits in IP addresses.

### Detailed Daemon Info
MooseFS daemons instructed by an `info` signal can now generate detailed info files in working directory, listing configuration, state, and module data for enhanced troubleshooting.

### CS-Master Consistency Checks
New mechanism to ensure consistency between master and chunk server data held in RAM.

### New `mfshdd.cfg` Options
Flags with limits and options for disks with chunks shared with other processes, allowing better calculation of available/occupied space on disks.

### New Offline Tool: `mfsmetasearch`
Allows offline searches for files in metadata based on various conditions such as chunk IDs, inodes, and attributes.

### Updated Logging System
Clear division of log messages by severity levels, with customizable logging options and syslog reporting.

### New ACL Handling Tools
Useful for operating systems without built-in ACL management tools.

### General Algorithm Improvements
Numerous improvements to algorithms for faster and more efficient MooseFS performance, especially in edge-case scenarios.

## Quality of Life Improvements

- **Faster Metadata Loading** at master server startup.
- **`mfssupervisor` Tool** for on-demand metadata saves.
- **`mfsfileinfo` Tool** can now display chunk paths on chunk servers.
- **Enhanced Chunk Rebalance Algorithms**.
- **More Configuration Options** for various operations and storage management.

## Community Contributions
During the development of this version, we received valuable feedback and suggestions from our community, many of which have been implemented. A huge thank you to everyone who contributed to shaping this release of MooseFS!

## MooseFS PRO
If you are looking for high availability (HA) for master servers, improved disk space management with enhanced data protection through EC configurations with higher redundancy levels, and dedicated support, we also offer MooseFS PRO version. This version is designed for enterprise environments where uptime and data integrity are critical, providing features like automatic failover for master servers, more advanced erasure coding for greater resilience against hardware failures, and professional support services to ensure smooth operation and rapid issue resolution. MooseFS PRO adds an extra layer of reliability and optimization for those managing large-scale, mission-critical storage systems.

---

We recommend to update your MooseFS clusters to this version.

Thank you,
Piotr / MooseFS Team

<hr /><em>This discussion was created from the release <a href='https://github.com/moosefs/moosefs/releases/tag/v4.56.5'>MooseFS 4.56.5</a>.</em>

**Comment by Lindsay-Mathieson** [comment]
Oh wow, this is really big!
Are there community repositories for it? Only seeing 3.x on the official MooseFS download page

**Reply by chogata** [reply]
Version 4 repository is now up and running.

**Comment by zhuaguoguo** [comment]
I am very concerned about this infomation and I want to upgrade the cluster. May I ask how to install the latest version? My system is centos8.

**Reply by selboo** [reply]
rpm/mfs.spec file in the repository, Can be packaged yourself

```
rpmbuild -bb rpm/mfs.spec
```

Or wait for the repository to update. http://repository.moosefs.com/moosefs-4

**Reply by Lindsay-Mathieson** [reply]
Thanks, I might have a play at building debs, but suspect the repo will be updated before I figure that out 😁

**Reply by zhuaguoguo** [reply]
thank you very much!

**Reply by chogata** [reply]
Version 4 repository is now up and running.

**Reply by Lindsay-Mathieson** [reply]
> Version 4 repository is now up and running.

Yup, thanks! have been trying it out for a few days now, very impressed.

**Comment by zhuaguoguo** [comment]
I would like to know how to set up EC4+1.

**Comment by Songdantes** [comment]
I would like to know how to set up EC4+1.

**Comment by xandrus** [comment]
Hi,

the simplest way to create EC4+1 storage class is to run this command:
```
cd /mnt/moosefs
mfsscadmin create -K 2 -A @4+1 -d 1 ec41
mfssetsclass -r ec41 /mnt/moosefs/my_ec41_folder
```
Let's break down `mfsscadmin create -K 2 -A @4+1 -d 1 ec41` line into parts:
`mfscadmin create` - create a new storage class

` -K ` mandatory parameter, that tells the system on which chunk servers, defined by the KEEP_LABELS expression, the chunk(s) should be kept always, except for special conditions like creating, archiving and deleting (moving to Trash), if defined.
It means to keep two copies on all available chunk servers.

`-A` optional parameter, that tells the system on which chunk servers, defined by the ARCH_LABELS expression, the chunk(s) should be kept for archiving  purposes. The system starts  to treat a chunk as archive, when atime/mtime/ctime (as set by -o) of the file it belongs to is older than the number of hours specified with -d option.

`@4+1` this is the expression that will enable erasure coding on chunk servers. EC labels expression starts with @ sign, followed by a number of data parts then + sign and a number that says how many parity parts the chunk should have.

`-d` set delay used to switch labels from 'keep' to 'archive' - if not specified default value 24[h] will be used.
In our case, we set the delay to one hour, so after one hour, the chunks will be migrated from keep state to archive state. 

It is worth mentioning that in MooseFS 4 only chunks in archive state can be stored in Erasure Codes, which means that write operations are always performed on standard "copies" but read operations can be performed on EC chunks.
Please definitely check the `man mfsscadmin`

**Reply by zhuaguoguo** [reply]
Thanks.
But after I execute this order "mfsscadmin create -K 2 -A @4+1 -d 1 ec41" in my mfsmount directory，it reports the following error:


parse error, next char: '@'
mfs storage class admin tool

usage:
        mfsscadmin [/mountpoint] create|make [-a admin_only] [-m mode] [-C create_labels] -K keep_labels [-A archive_labels -d archive_delay] sclass [sclass ...]
        mfsscadmin [/mountpoint] create|make [-a admin_only] [-m mode] LABELS sclass [sclass ...]
        mfsscadmin [/mountpoint] modify|change [-f] [-a admin_only] [-m mode] [-C create_labels] [-K keep_labels] [-A archive_labels] [-d archive_delay] sclass [sclass ...]
        mfsscadmin [/mountpoint] delete|remove sclass [sclass ...]
        mfsscadmin [/mountpoint] copy|duplicate src_sclass dst_sclass [dst_sclass ...]
        mfsscadmin [/mountpoint] rename src_sclass_name dst_sclass_name
        mfsscadmin [/mountpoint] list [-l]

create/modify options:
 -m - set mode (options are: 'l' for 'loose', 's' for 'strict' and 'd' or not specified for 'default')
    'default' mode: if there are overloaded servers, system will wait for them, but in case of no space available will use other servers (disregarding the labels).
    'strict' mode: the system will wait for overloaded servers, but when there is no space on servers with correct labels then will not use them (return ENOSPC or leave chunk as unde
rgoal).
    'loose' mode: the system will disregard the labels in both cases.
 -a - set admin only mode ( 0 - anybody can use this storage class, 1 - only admin can use this storage class ) - by default it set to 0
 -C - set labels used for creation chunks - when not specified then 'keep' labels are used
 -K - set labels used for keeping chunks - must be specified
 -A - set labels used for archiving chunks - when not specified then 'keep' labels are used
 -d - set number of days used to switch labels from 'keep' to 'archive' - must be specified when '-A' option is given
 -f - force modification of classes 1 to 9 (modify only)

list options:
 -l - lists storage classes with definitions (long format)

If '/mountpoint' parameter is not specified then current directory is used (it might be any subfolder of mountpoint).
All actions but list need 'admin' flag specified in mfsexports.cfg

**Reply by zhuaguoguo** [reply]
Acturally, I can make sure I am in the mount directory and I also have admin permisson.

**Reply by Songdantes** [reply]
> Acturally, I can make sure I am in the mount directory and I also have admin permisson.

the mfsscadmin and mfsmount all have to the version moosefs 4

**Reply by Songdantes** [reply]
> Hi,
> 
> the simplest way to create EC4+1 storage class is to run this command:
> 
> ```
> cd /mnt/moosefs
> mfsscadmin create -K 2 -A @4+1 -d 1 ec41
> mfssetsclass -r ec41 /mnt/moosefs/my_ec41_folder
> ```
> 
> Let's break down `mfsscadmin create -K 2 -A @4+1 -d 1 ec41` line into parts: `mfscadmin create` - create a new storage class
> 
> `-K` mandatory parameter, that tells the system on which chunk servers, defined by the KEEP_LABELS expression, the chunk(s) should be kept always, except for special conditions like creating, archiving and deleting (moving to Trash), if defined. It means to keep two copies on all available chunk servers.
> 
> `-A` optional parameter, that tells the system on which chunk servers, defined by the ARCH_LABELS expression, the chunk(s) should be kept for archiving purposes. The system starts to treat a chunk as archive, when atime/mtime/ctime (as set by -o) of the file it belongs to is older than the number of hours specified with -d option.
> 
> `@4+1` this is the expression that will enable erasure coding on chunk servers. EC labels expression starts with @ sign, followed by a number of data parts then + sign and a number that says how many parity parts the chunk should have.
> 
> `-d` set delay used to switch labels from 'keep' to 'archive' - if not specified default value 24[h] will be used. In our case, we set the delay to one hour, so after one hour, the chunks will be migrated from keep state to archive state.
> 
> It is worth mentioning that in MooseFS 4 chunks only in archive mode can be stored in Erasure Codes, which means that write operations are always performed on standard "copies" but read operations can be performed on EC chunks. Please definitely check the `man mfsscadmin`

<img width="1435" alt="image" src="https://github.com/user-attachments/assets/f4be0c2a-5951-4645-9f36-20bb7678a055">

When I  run this command ,  after  one day , the EC chunks is 0% and Space saved
by EC is not yet , anyting else to do?  ( in the mfsmaster.cfg file i set DEFAULT_EC_DATA_PARTS = 4)

**Reply by chogata** [reply]
@Songdantes 
First, the default archive delay is based on ctime, so if anything is modifying your files' data OR metadata, they won't revert to ARCHIVE state and convert to EC format.
Second, the delay of 1 day means, that MooseFS will start to turn on the archive flag during the first file loop that happens after the files haven't been modified for a full day, and then the chunk loop has to convert the chunks to the EC format. Depending on how long the loops take on your system, it may add anything from 10 minutes to a coupl of hours to the delay.
You wrote your post 2 days ago, has anything changed since then? If not, can you post the screenshot of your "Resources" tab?

**Reply by Songdantes** [reply]
The file   I wrote use this shell "dd if=/dev/zero of="/mfs_mount/$filename" bs=2M count=1 status=none" ，I don't know if it's related to this.   the EC chunks is still 0% and Space saved
by EC is not yet.  This is the "Resources"
<img width="1440" alt="image" src="https://github.com/user-attachments/assets/5c72ef1b-ea00-4b36-bbff-3c516a907605">

**Reply by chogata** [reply]
And we have the answer: the table shows "EC KEEP ONLY", which means you don't have enough chunk servers connected to use EC. The minimum required number for EC 4+1 is 6 chunk servers, for EC 8+1 is 10 chunk servers. If you have 5 (9 for EC8), MooseFS will keep the existing EC chunks, but not create new chunks in this format. If you have 4 or less (8 or less for EC8), MooseFS will revert all data to KEEP (copy) format.

This is because each part of EC needs to be kept on a different chunk server. 4+1 means 5 parts, which means 5 servers are necessary to at least keep it. 6 are required to convert new chunks to EC, because we want the system to still work (aka have a place to replicate a missing part) if one server fails.

**Reply by Lindsay-Mathieson** [reply]
> This is because each part of EC needs to be kept on a different chunk server. 4+1 means 5 parts, which means 5 servers are necessary to at least keep it. 6 are required to convert new chunks to EC

Does having mismatched chunkserver sizes impact that? I currently have 6 chunkservers
* CS1: 8TB
* CS2: 8TB
* CS3: 6TB
* CS4: 6TB
* CS5: 16TB
* CS6: 16TB

Currently at Goal 2, 43% used. Would converting it to EC 4+1 cause it to die a horrible death? 😁

**Reply by Lindsay-Mathieson** [reply]
> It is worth mentioning that in MooseFS 4 chunks only in archive mode can be stored in Erasure Codes, which means that write operations are always performed on standard "copies" but read operations can be performed on EC chunks

Interesting, is there any particular reason for that? simplifies the write process?

Does that mean when an archived (ec) file is written to, the portion that is changed is written as R2, or is the entire file converted to R2 (or whatever -K goal is set)

**Reply by chogata** [reply]
When a file is written to, the chunk (or chunks) that are modified are first converted back to copies, to whatever definition (number of copies and optionally labels) you have for you KEEP state. Then the actual i/o happens. It's because of the nature of the EC format, which is unsuitable for modifications. Chunks that aren't modified aren't converted - they stay in EC. And chunks that have been modified will be converted back to EC eventually - when depends on the exact definition of the storage class used.

You can use EC on "mismatched" chunk servers, MooseFS will manage just fine. If you try to convert all your data to EC at the same time, you may notice an increase in CPU usage on your hardware, but default replication limits are not high, so all should be fine.

**Reply by Lindsay-Mathieson** [reply]
Thanks! Makes sense. Well suited to my use case which is media storage. Testing ec4+1 on a subsection of data and it seems to be converting smoothly.

I do love how flexible MooseFS is for mixing storage classes and file.

**Reply by Lindsay-Mathieson** [reply]
> . If you try to convert all your data to EC at the same time, you may notice an increase in CPU usage on your hardware, but default replication limits are not high

I currently have the rep limits cranked way up 

`CHUNKS_WRITE_REP_LIMIT = 2,1,1,16,16`
`CHUNKS_READ_REP_LIMIT = 10,5,2,20,20`

As I recently added the 2 16TB disks and wanted to speed up the rebalance - its thrashing the disks :) CPU usage is minimal though.

**Reply by chogata** [reply]
It all depends on what you use - some people use NVMe disks and then the CPU goes up. On old fashioned rotational disks, or even older SSDs, the disk will be the slowest component.

**Comment by vaab** [comment]
Also interested to try new EC storage feature with previously given example, but it fails:
```
$ mfsscadmin create -K 2 -A @4+1 -d 1 ec41
storage class make ec41: error: Incompatible version
```
With freshly baked debian package install on v4.56.6 (commit 5afd7b5), using `debuild -i -us -uc -b` (running on Ubuntu 22.04). Double checked:
- `mfsmaster`, `mfsmount` (re-mounted with new version), `mfsscadmin` process were all replaced where necessary and double checked. I'm on my `/mnt/mfs` directory when issuing the command.
- the mount export is correctly flagged with `admin`.
```
$ dpkg -l moosefs-*
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name            Version          Architecture Description
+++-===============-================-============-=========================================================
ii  moosefs-cgi     4.56.6-1         amd64        MooseFS CGI Monitor
ii  moosefs-cgiserv 4.56.6-1         amd64        Simple CGI-capable HTTP server to run MooseFS CGI Monitor
ii  moosefs-cli     4.56.6-1         amd64        MooseFS CLI Utility
ii  moosefs-client  4.56.6-1         amd64        MooseFS client tools
rc  moosefs-common  3.0.116-1ubuntu2 all          MooseFS - common files
ii  moosefs-master  4.56.6-1         amd64        MooseFS master (metadata) server
```
All 6 chunkservers are also migrated, and the new UI is gorgeous and everything seems to go well after migration from a `3.0.118`.
I guess a more explicit error message would help ?

**Reply by chogata** [reply]
What about the chunk servers? I don't see them in the listing (probably because it's from the master machine), are they all upgraded to 4.56.6? Can you post your "Servers" and "Mounts" screenshots, or output from `mfscli -H your.masterhost -SCS -SMS` (substitute "your.masterhost" with your instances' master host).

**Reply by vaab** [reply]
As stated "all 6 chunkservers are also migrated". 
![image](https://github.com/user-attachments/assets/b1bca29a-decd-4205-87e5-e445a84cda01)

But I had a mount from another client still in `3.0.118`, that I didn't think was important (it was not the one I was using to run the command). After umounting it I could run the `mfscsadmin create -K 2 -A @4+1 -d 1 ec41` command with no issue. Many thanks, and sorry for my oversight on this point.

I still feel a more explicit error message would help ;-).

**Reply by chogata** [reply]
I'm happy it's now solved and we will think about expanding the error message :)

**Comment by yaroslav-gwit** [comment]
First of all, thank you for all your hard work on this project.

You mentioned that the HA code would become a part of the public repo in v4:
https://github.com/moosefs/moosefs/issues/85#issuecomment-390521097

I tried a quick search for `shadow` or `ucarp` (similar to how it's implemented in LizardFS) but it didn't return anything interesting. Was the decision to publish the HA functionality "rolled back", or am I looking for a wrong thing altogether?

**Reply by pkonopelko** [reply]
I believe the answer for your question is here: https://github.com/moosefs/moosefs/issues/39#issuecomment-2385496411

**Reply by yaroslav-gwit** [reply]
Thank you for the explanation, that makes sense.
